cartridge/upload.lua

```diff
@@ -0,0 +1,322 @@
+            if retmap == nil or retmap[uri] == nil then
+                local err = errmap and errmap[uri]
+                if err == nil then
+                    err = UploadError:new('Unknown error at %s', uri)
+                end
+                log.error('Error uploading files to %s:\n%s', uri, err)
+                _upload_error = err
+            elseif retmap[uri] == true then
+                table.insert(transmitters_list, uri)
+            end
+        end
+
+        if _upload_error ~= nil then
+            goto cleanup
+        end
+    end
+
+    do -- transmit
+        local retmap, errmap = pool.map_call(
+            '_G.__cartridge_upload_transmit', {upload_id, payload},
+            {
+                uri_list = transmitters_list,
+                timeout = opts.transmission_timeout,
+            }
+        )
+
+        for _, uri in ipairs(transmitters_list) do
+            if retmap == nil or retmap[uri] == nil then
+                local err = errmap and errmap[uri]
+                if err == nil then
+                    err = UploadError:new('Unknown error at %s', uri)
+                end
+                log.error('Error transmitting data to %s:\n%s', uri, err)
+                _upload_error = err
+            end
+        end
+
+        if _upload_error ~= nil then
+            goto cleanup
+        end
+    end
+
+    do -- finish
+        local retmap, errmap = pool.map_call(
+            '_G.__cartridge_upload_finish', {upload_id},
+            {
+                uri_list = opts.uri_list,
+                timeout = opts.netbox_call_timeout,
+            }
+        )
+
+        for _, uri in ipairs(opts.uri_list) do
+            if retmap == nil or retmap[uri] == nil then
+                log.warn(
+                    'Error finishing upload on %s:\n%s',
+                    uri, errmap and errmap[uri]
+                )
+            end
+        end
+    end
+
+::cleanup::
+    do -- cleanup
+        local retmap, errmap = pool.map_call(
+            '_G.__cartridge_upload_cleanup', {upload_id},
+            {
+                uri_list = opts.uri_list,
+                timeout = opts.netbox_call_timeout,
+            }
+        )
+
+        for _, uri in ipairs(opts.uri_list) do
+            if retmap == nil or retmap[uri] == nil then
+                log.warn(
+                    'Error cleaning up %s:\n%s',
+                    uri, errmap and errmap[uri]
+                )
+            end
+        end
+    end
+
+    if _upload_error ~= nil then
+        return nil, _upload_error
+    end
+
+    return upload_id
+end
+
+_G.__cartridge_upload_begin = function(...) return errors.pcall('E', upload_begin, ...) end
+_G.__cartridge_upload_transmit = function(...) return errors.pcall('E', upload_transmit, ...) end
+_G.__cartridge_upload_finish = function(...) return errors.pcall('E', upload_finish, ...) end
+_G.__cartridge_upload_cleanup = function(...) return errors.pcall('E', upload_cleanup, ...) end
```

Comment:
You don't actually need to pass `...` here because you exactly know a number of arguments - one upload_id.
varargs are bit slower than usual function call with determined number of arguments.

Reply:
I'm ok to sacrifice that performance for the sake of code brevity. Hope it's not too much.